File: 20_杂题/atc競プロ/AtCoder Beginner Contest/291/f.py
import sys
from typing import List

sys.setrecursionlimit(int(1e9))
input = lambda: sys.stdin.readline().rstrip("\r\n")
MOD = 998244353
INF = int(4e18)
# N 個の都市があり、都市
# 1, 都市
# 2,
# …, 都市
# N と番号づけられています。
# いくつかの異なる都市の間は一方通行のテレポーターによって移動できます。 都市
# i
# (1≤i≤N) からテレポーターによって直接移動できる都市は 0 と 1 からなる長さ
# M の文字列
# S
# i
# ​
#   によって表されます。具体的には、
# 1≤j≤N に対して、

# 1≤j−i≤M かつ
# S
# i
# ​
#   の
# (j−i) 文字目が 1 ならば、都市
# i から都市
# j に直接移動できる。
# そうでない時、都市
# i から都市
# j へは直接移動できない。
# k=2,3,…,N−1 に対して次の問題を解いてください。

# テレポータを繰り返し使用することによって、都市
# k を通らずに都市
# 1 から 都市
# N へ移動できるか判定し、 できるならばそのために必要なテレポーターの使用回数の最小値を、 できないならば
# −1 を出力せよ。
if __name__ == "__main__":
    n, m = map(int, input().split())
    words = [input() for _ in range(n)]
    adjList = [set() for _ in range(n)]
    rAdjList = [set() for _ in range(n)]
    for i, w in enumerate(words):
        for j, c in enumerate(w):
            if c == "1":
                adjList[i].add(i + j + 1)
                rAdjList[i + j + 1].add(i)

    from collections import deque

    def bfs(start: int, adjList: List[List[int]]) -> List[int]:
        n = len(adjList)
        dist = [INF] * n
        dist[start] = 0
        queue = deque([start])
        while queue:
            cur = queue.popleft()
            for next in adjList[cur]:
                cand = dist[cur] + 1
                if cand < dist[next]:
                    dist[next] = cand
                    queue.append(next)
        return dist

    # 枚举他左右范围即可
    dist1 = bfs(0, adjList)
    dist2 = bfs(n - 1, rAdjList)

    for i in range(1, n - 1):
        # 枚举左右落脚点(10个*10个)
        res = INF
        for left in range(max(0, i - m), i):
            for right in range(i + 1, min(n, i + m + 1)):
                if right in adjList[left]:
                    cand = dist1[left] + dist2[right] + 1
                    res = min(res, cand)
        print(-1 if res == INF else res)

File: tmp/abc338/f.go
package main

import (
	"bufio"
	"fmt"
	stdio "io"
	"os"
	"strconv"
)

var io *Iost

type Iost struct {
	Scanner *bufio.Scanner
	Writer  *bufio.Writer
}

func NewIost(fp stdio.Reader, wfp stdio.Writer) *Iost {
	const BufSize = 2000005
	scanner := bufio.NewScanner(fp)
	scanner.Split(bufio.ScanWords)
	scanner.Buffer(make([]byte, BufSize), BufSize)
	return &Iost{Scanner: scanner, Writer: bufio.NewWriter(wfp)}
}
func (io *Iost) Text() string {
	if !io.Scanner.Scan() {
		panic("scan failed")
	}
	return io.Scanner.Text()
}
func (io *Iost) Atoi(s string) int                 { x, _ := strconv.Atoi(s); return x }
func (io *Iost) Atoi64(s string) int64             { x, _ := strconv.ParseInt(s, 10, 64); return x }
func (io *Iost) Atof64(s string) float64           { x, _ := strconv.ParseFloat(s, 64); return x }
func (io *Iost) NextInt() int                      { return io.Atoi(io.Text()) }
func (io *Iost) NextInt64() int64                  { return io.Atoi64(io.Text()) }
func (io *Iost) NextFloat64() float64              { return io.Atof64(io.Text()) }
func (io *Iost) Print(x ...interface{})            { fmt.Fprint(io.Writer, x...) }
func (io *Iost) Printf(s string, x ...interface{}) { fmt.Fprintf(io.Writer, s, x...) }
func (io *Iost) Println(x ...interface{})          { fmt.Fprintln(io.Writer, x...) }

const INF int = 1e18

func main() {
	in := os.Stdin
	out := os.Stdout
	io = NewIost(in, out)
	defer func() {
		io.Writer.Flush()
	}()

	n, m := io.NextInt(), io.NextInt()
	edges := make([][3]int, m)
	for i := 0; i < m; i++ {
		edges[i][0], edges[i][1], edges[i][2] = io.NextInt(), io.NextInt(), io.NextInt()
		edges[i][0]--
		edges[i][1]--
	}
	adjList := make([][][2]int, n)
	for _, edge := range edges {
		adjList[edge[0]] = append(adjList[edge[0]], [2]int{edge[1], edge[2]})
	}

	target := (1 << n) - 1
	dist := make([][]int, n)
	for i := range dist {
		dist[i] = make([]int, 1<<n)
		for j := range dist[i] {
			dist[i][j] = INF
		}
	}

	queue := NewHeap(func(a, b H) bool { return a[2] < b[2] }, nil)
	for i := 0; i < n; i++ {
		queue.Push([3]int{i, 1 << i, 0})
		dist[i][1<<i] = 0
	}
	for queue.Len() > 0 {
		cur := queue.Pop()
		curPoint, curState, curCost := cur[0], cur[1], cur[2]
		for _, next := range adjList[curPoint] {
			nextPoint, nextCost := next[0], next[1]
			nextState := curState | (1 << nextPoint)
			if dist[nextPoint][nextState] > curCost+nextCost {
				dist[nextPoint][nextState] = curCost + nextCost
				queue.Push([3]int{nextPoint, nextState, curCost + nextCost})
			}
		}
	}

	res := INF

	for i := 0; i < n; i++ {
		res = min(res, dist[i][target])
	}
	if res != INF {
		io.Println(res)
		return
	}
	io.Println("No")

}

type H = [3]int

func NewHeap(less func(a, b H) bool, nums []H) *Heap {
	nums = append(nums[:0:0], nums...)
	heap := &Heap{less: less, data: nums}
	heap.heapify()
	return heap
}

type Heap struct {
	data []H
	less func(a, b H) bool
}

func (h *Heap) Push(value H) {
	h.data = append(h.data, value)
	h.pushUp(h.Len() - 1)
}

func (h *Heap) Pop() (value H) {
	if h.Len() == 0 {
		panic("heap is empty")
	}

	value = h.data[0]
	h.data[0] = h.data[h.Len()-1]
	h.data = h.data[:h.Len()-1]
	h.pushDown(0)
	return
}

func (h *Heap) Top() (value H) {
	value = h.data[0]
	return
}

func (h *Heap) Len() int { return len(h.data) }

func (h *Heap) heapify() {
	n := h.Len()
	for i := (n >> 1) - 1; i > -1; i-- {
		h.pushDown(i)
	}
}

func (h *Heap) pushUp(root int) {
	for parent := (root - 1) >> 1; parent >= 0 && h.less(h.data[root], h.data[parent]); parent = (root - 1) >> 1 {
		h.data[root], h.data[parent] = h.data[parent], h.data[root]
		root = parent
	}
}

func (h *Heap) pushDown(root int) {
	n := h.Len()
	for left := (root<<1 + 1); left < n; left = (root<<1 + 1) {
		right := left + 1
		minIndex := root

		if h.less(h.data[left], h.data[minIndex]) {
			minIndex = left
		}

		if right < n && h.less(h.data[right], h.data[minIndex]) {
			minIndex = right
		}

		if minIndex == root {
			return
		}

		h.data[root], h.data[minIndex] = h.data[minIndex], h.data[root]
		root = minIndex
	}
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func mins(nums ...int) int {
	res := nums[0]
	for _, num := range nums {
		if num < res {
			res = num
		}
	}
	return res
}

func maxs(nums ...int) int {
	res := nums[0]
	for _, num := range nums {
		if num > res {
			res = num
		}
	}
	return res
}
